### ISSUE TITLE
parse_filetime() not handling ValueErrors for timestamps out of range

### ISSUE THREAD

**Issue by hiddenillusion** [williballenthin/INDXParse#16]
If BinaryParser's parse_filetime() doesn't check for a ValueError it will throw an exception and halt parsing. A few $MFT records I'm looking at have some weird timestamp values. Here's what the qwords for one of the records looks like prior to attempting to return it as a Windows 64bit timestamp:

130380829518672080 -> SI Modified
130380829518672080 -> SI Accessed
130380829518672080 -> SI Created
111595705755238660 -> SI Changed, blank in EnCase (Entry Modified)

Traceback:
```
	'si_changed': si.changed_time(),
	File "BinaryParser.py", line 521, in basic_no_length_handler
	return f(offset)
	File "BinaryParser.py", line 875, in unpack_filetime
	return parse_filetime(self.unpack_qword(offset))
	File "BinaryParser.py", line 274, in parse_filetime
	return datetime.utcfromtimestamp(float(qword) * 1e-7 - 11644473600)
ValueError: timestamp out of range for platform localtime()/gmtime() function
```

Proposed solution is to except the ValueError and return a standard EPOCH datetime that way there's still a timestamp associated with that attribute vs. just ignoring it:

```
def parse_filetime(qword):
    # see http://integriography.wordpress.com/2010/01/16/using-phython-to-parse-and-present-windows-64-bit-timestamps/
    if qword:
        try:
            return datetime.utcfromtimestamp(float(qword) * 1e-7 - 11644473600)
        except ValueError:       
            return datetime(1970, 1, 1, 0, 0, 0)
```

**Comment by ajnelson** [comment]
I think it would be sensible to give a default timestamp.  It seems the default throughout the rest of the code is 1970.  It may hide data or code problems, though, if the error is silent.

**Comment by hiddenillusion** [comment]
Agreed - I figured putting the default timestamp would atleast stop the
processing from exiting and/or not silently drop the record and it's
details depending on how it's being called. I've seen this method being
used a lot these days and I'm a fan. I cross validated on other tools and
they either raise the issue and exit, silently drop the entire record or
just display a blank timestamp (e.g. - EnCase).

I think supplementing the error checking with logging is definitely ideal -
I use the same method so I can always see the issues, but I didn't want to
bloat the code on the first iteration of bringing up the issue.

I know I've seen the g verbose (quick, hackish logging) method being used
by willi before as a lighter logger but if true logging  like you've
mentioned is the route that willi thinks is best I'd be happy to implement
and give a PR.

- Glenn
@hiddenillusion

:: Sent from my mobile device, please excuse any typos ::

**Comment by williballenthin** [comment]
`g_verbose` is an artifact of a long time ago when I had a weaker grasp on the Python ecosystem. Using `logging` is absolutely the correct approach. As noted elsewhere, I'll be fixing this (and adding a separate issue).

@hiddenillusion @ajnelson I propose logging a WARNING level entry when the default timestamp is picked. What do you guys think?

The more correct approach is probably to set the field to `None`, and handle this during output, but it make processing (such as sorting) and formatting (during output) much messier.  Therefore, I think probably any underlying libraries (such as `python-ntfs`) should return exactly the data present, or throw an error. However, perhaps its reasonable for this data carving script to emphasize the algorithm?

**Comment by williballenthin** [comment]
should be closed in 67ef5e7753e5c7263553e32971c14e016c9f83cc. please reopen if there's still an issue.
